services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/HankeServiceImpl.kt

```diff
@@ -178,11 +180,35 @@ open class HankeServiceImpl(
     }
 
     @Transactional
-    override fun deleteHanke(hanke: Hanke, userId: String) {
-        hankeRepository.deleteById(hanke.id!!)
-        hankeLoggingService.logDelete(hanke, userId)
+    override fun deleteHanke(hankeId: Int, userId: String): Boolean {
+        hankeRepository.findByIdOrNull(hankeId)?.let { entity ->
+            if (entity.anyHakemusProcessingInAllu()) {
+                throw HankeAlluConflictException(
+                    "Hanke ${entity.hankeTunnus} has hakemus in Allu processing. Cannot delete."
+                )
+            }
+            hankeRepository.deleteById(hankeId)
+            hankeLoggingService.logDelete(createHankeDomainObjectFromEntity(entity), userId)
+            return true
+        }
+        return false
     }
 
+    /**
+     * An application is being processed in Allu if status it is NOT pending anymore. Pending status
+     * needs verification from Allu to prevent race condition.
+     */
+    private fun HankeEntity.anyHakemusProcessingInAllu(): Boolean =
+        hakemukset.any {
+            logger.info { "Hakemus ${it.id} has alluStatus ${it.alluStatus}" }
+            when (it.alluStatus) {
+                null,
+                ApplicationStatus.PENDING,
+                ApplicationStatus.PENDING_CLIENT -> !applicationService.isStillPending(it.alluid)
+                else -> true
+            }
```

Comment:
This could be moved to an overloaded isStillPending method that takes ApplicationEntity as a parameter. This way others can use the same logic.

The documentation could say that applications can't move back to a pending status after leaving it. To justify why the else-branch is valid.

Reply:
Done.

Reply:
Done.